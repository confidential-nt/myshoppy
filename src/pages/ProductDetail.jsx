import React from "react";
import { useLocation, useParams } from "react-router-dom";
import { useUserContext } from "../context/UserContext";
import { useUserRepositoryContext } from "../context/UserRepositoryContext";

export default function ProductDetail() {
  const {
    state: { product },
  } = useLocation();

  const { uid } = useUserContext();
  const { userRepository } = useUserRepositoryContext();

  const { productId } = useParams();

  const handleClick = (e) => {
    if (!uid) return;
  };

  return (
    <div className="pl-5 pr-5">
      <strong className="inline-block mt-3 mb-3 pl-4 font-normal text-xs text-gray-500">
        &gt; {product.category}
      </strong>
      <div className="flex">
        <img src={product.imageURL} alt={product.name} className="basis-3/5" />
        <div className="basis-2/5 ml-5 pt-3">
          <div className="border-b pb-1 mb-2">
            <h3 className="font-semibold text-xl mb-1">{product.name}</h3>
            <span className="font-semibold">₩{product.price}</span>
          </div>
          <p className="text-sm mb-3">{product.desc}</p>
          <div className="flex items-center mb-3">
            <span className="text-shoppypink text-xs mr-2">옵션:</span>
            <select
              name="options"
              className="block grow border border-dashed border-shoppypink pl-1 pr-1 pt-1 pb-1"
            >
              {product.options.split(",").map((s) => (
                <option key={s} value={s}>
                  {s}
                </option>
              ))}
            </select>
          </div>
          <button
            type="button"
            className="block cursor-pointer relative bg-shoppypink w-5/6 mr-auto ml-auto text-white text-xl pt-1 pb-1 after:content-[''] after:w-full after:h-full after:absolute after:top-0 after:left-0 after:bg-black/30 after:hidden hover:after:block  disabled:cursor-not-allowed disabled:hover:after:hidden disabled:bg-shoppypink/30"
            disabled={!Boolean(uid)}
            onClick={handleClick}
          >
            장바구니에 추가
          </button>
        </div>
      </div>
    </div>
  );
}
