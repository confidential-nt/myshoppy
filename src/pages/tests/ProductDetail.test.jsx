import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import {
  withAllUserContexts,
  withReactQuery,
  withRouter,
} from "../../tests/utils";
import { Route } from "react-router-dom";
import ProductDetail from "../ProductDetail";
import { fakeProduct } from "../../tests/products";

describe("ProductDetail", () => {
  const userRepository = {
    updateCartsOnce: jest.fn(),
  };

  afterEach(() => {
    userRepository.updateCartsOnce.mockReset();
  });

  it("스냅샷 테스트", () => {
    const { asFragment } = renderProductDetail(null);
    expect(asFragment()).toMatchSnapshot();
  });

  it("로그인하지 않은 상태면 장바구니 추가 버튼이 disabled 상태여야한다", () => {
    renderProductDetail(null);
    const button = screen.getByRole("button", {
      name: "장바구니에 추가",
    });
    expect(button).toBeDisabled();
  });

  it("로그인한 상태면 장바구니 추가 버튼이 abled 상태여야한다", () => {
    renderProductDetail("1");
    const button = screen.getByRole("button", {
      name: "장바구니에 추가",
    });
    expect(button).not.toBeDisabled();
  });

  it("장바구니에 아이템이 추가되면 '장바구니에 추가되었습니다'라는 메시지가 떠야한다", async () => {
    renderProductDetail("1");
    const button = screen.getByRole("button", {
      name: "장바구니에 추가",
    });
    await userEvent.click(button);
    await waitFor(() =>
      expect(
        screen.getByText("✔️ 장바구니에 추가되었습니다.")
      ).toBeInTheDocument()
    );
  });

  function renderProductDetail(uid) {
    return render(
      withAllUserContexts(
        withReactQuery(
          withRouter(<Route element={<ProductDetail />} path="/products/1" />, {
            pathname: "/products/1",
            state: { product: fakeProduct },
            key: "fake-key",
          })
        ),
        userRepository,
        {
          uid,
          logUserIn: () => {},
        }
      )
    );
  }
});
