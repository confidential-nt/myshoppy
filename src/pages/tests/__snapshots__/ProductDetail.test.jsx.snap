// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductDetail 스냅샷 테스트 1`] = `
<DocumentFragment>
  <div
    class="pl-5 pr-5"
  >
    <strong
      class="inline-block mt-3 mb-3 pl-4 font-normal text-xs text-gray-500"
    >
      &gt; 여성
    </strong>
    <div
      class="flex"
    >
      <img
        alt="핫핑크 후드 HOODY"
        class="basis-3/5"
        src="http://image/"
      />
      <div
        class="basis-2/5 ml-5 pt-3"
      >
        <div
          class="border-b pb-1 mb-2"
        >
          <h3
            class="font-semibold text-xl mb-1"
          >
            핫핑크 후드 HOODY
          </h3>
          <span
            class="font-semibold"
          >
            ₩40000
          </span>
        </div>
        <p
          class="text-sm mb-3"
        >
          스타일리쉬한 핑크색이 돋보이는 핫핑크색 후드
        </p>
        <div
          class="flex items-center mb-3"
        >
          <span
            class="text-shoppypink text-xs mr-2"
          >
            옵션:
          </span>
          <select
            class="block grow border border-dashed border-shoppypink pl-1 pr-1 pt-1 pb-1"
            name="options"
          >
            <option
              value="xs"
            >
              xs
            </option>
            <option
              value="s"
            >
              s
            </option>
            <option
              value="m"
            >
              m
            </option>
            <option
              value="l"
            >
              l
            </option>
            <option
              value="xl"
            >
              xl
            </option>
          </select>
        </div>
        <button
          class="block cursor-pointer relative bg-shoppypink w-5/6 mr-auto ml-auto text-white text-xl pt-1 pb-1 after:content-[''] after:w-full after:h-full after:absolute after:top-0 after:left-0 after:bg-black/30 after:hidden hover:after:block  disabled:cursor-not-allowed disabled:hover:after:hidden disabled:bg-shoppypink/30"
          disabled=""
          type="button"
        >
          장바구니에 추가
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
